File: 142_Triangle.py
"""
142. Triangle
Bottom-up DP collapsing rows to compute min path sum.
"""
from typing import List

def minimum_total(triangle: List[List[int]]) -> int:
    dp = triangle[-1][:]
    for layer in range(len(triangle)-2, -1, -1):
        for i in range(layer+1):
            dp[i] = triangle[layer][i] + min(dp[i], dp[i+1])
    return dp[0]

if __name__ == '__main__':
    print('minimumTotal =', minimum_total([[2],[3,4],[6,5,7],[4,1,8,3]]))

# Time Complexity: O(n^2)
# Space Complexity: O(n)

File: 142_Triangle.cpp
/*
142. Triangle
Find minimum path sum from top to bottom of a triangle. Use bottom-up DP to reuse space.
*/
#include <bits/stdc++.h>
using namespace std;

int minimumTotal(vector<vector<int>>& triangle) {
    int n = triangle.size();
    vector<int> dp(triangle.back());
    for (int layer = n-2; layer >= 0; --layer) {
        for (int i = 0; i <= layer; ++i) dp[i] = triangle[layer][i] + min(dp[i], dp[i+1]);
    }
    return dp[0];
}

int main() {
    vector<vector<int>> tri{{2},{3,4},{6,5,7},{4,1,8,3}};
    cout << "minimumTotal = " << minimumTotal(tri) << "\n"; // 11
    return 0;
}

// Time Complexity: O(n^2)
// Space Complexity: O(n)
